### ISSUE TITLE
RFC: Hydrodynamics

### ISSUE THREAD

**Issue by tttppp** [andybalaam/rabbit-escape#298]
This isn't Minecraft, but maybe it could get a little closer...

This is a discussion of water, how it affects rabbits and how it interacts with its surroundings.

I propose four water related states:

1. Springs - water is generated by these.
2. Water fall - This is a square through which water is falling.
3. River - This is water about half a square deep, which is non-fatal to rabbits.
4. Pool - This is water a whole square deep and is fatal to normal rabbits (introduction of snorkels may change things).

Water will follow a path according to the following rules:

1. A space with a water fall, spring or river above it becomes a water fall.
2. A water fall with ground or pool under it becomes a river.
3. A space with ground or pool under it and river next to it becomes a river.
4. A horizontal row of rivers squares bounded at both ends by walls/ramps becomes a horizontal row of pool squares.
5. A pool or river without ground under it becomes a water fall (e.g. could be caused by bashing).
6. A pool with water fall next to it becomes a river.
7. A horizontal collection of river tiles without a water fall or spring above them dries up (I think this rule in particular needs some thinking about).

Only solid blocks and ramps affect water - bridges have no impact.

Water movement/changes could either be at some speed, or instantaneous. I'm not sure how I feel about this.

**Comment by tttppp** [comment]
Ideas for puzzles this would create:

* Make an exit for water to leave through before bashing into the side of a pool.
* Get some rabbits from left chamber to middle chamber and then redirect water into left chamber, to allow safe passage into right chamber.
* Make a hole in the floor and then bridge over it to prevent the current chamber from filling up.
* Race against impending drowning to escape. (Assuming non-instantaneous water flow)

Something else I just thought of is that rivers would need to be able to run down ramps.

**Comment by andybalaam** [comment]
Definite puzzletunities but some serious coding and design to be done. Prototypes welcome.

**Comment by colonelfazackerley** [comment]
It will be fun. I think it will need particles[1] instead of sprites. I think the water in each cell will need numbers for velocity and how full the cell is.

[1]
https://en.wikipedia.org/wiki/Particle_system

**Comment by tttppp** [comment]
I totally agree that this would be a lot of coding (and testing!) but the main thing I wanted to agree short term was whether this was a good enough reason to reserve "water" as distinct from the static death squares.

I had a quick go at putting together a python-based water flow state machine today. The intention was to test the rules I put above, but it was not trivial.

I like the idea of modelling water with velocity and amount. My original intention was to simplify things so that water doesn't actually move, rather it 'infects' neighbouring squares in a "Game of Life" style.

**Comment by tttppp** [comment]
Something else that occurred to me was that even more puzzles could be created with this if we also had unbreakable walls (metal?) and Portal-style button/doors (i.e. step on a button to open a metal door). Water could pres buttons, meaning that a player might want to carefully redirect the water to a particular square.

</ramble>

**Comment by colonelfazackerley** [comment]
Metal blocks are still on the TODO list that pre-dates use of github issues. I like buttons too.

I fear we may be engaging in too much chat and not enough code.

**Comment by andybalaam** [comment]
Minecraft demonstrates that a cellular automata style of water flow can work. One thing I like about the proposed scheme is that water would interact only with blocks (except when it kills rabbits) so it would not complicate other code.

**Comment by andybalaam** [comment]
@GamingInky and I are hoping to do metal blocks so please don't start them

**Comment by andybalaam** [comment]
@tttppp post gifs of prototypes for fun.

**Comment by tttppp** [comment]
Prototype of water flow using Python:

https://ideone.com/Unvniz

Mini-example (~ is water, N is pool, other symbols you can look up in the script if you want to):
```
  x  
#  / 
#####

  *  
#  / 
#####

  *  
# |/ 
#####

  *  
# ~/ 
#####

  *  
#~~} 
#####

  *  
#NN] 
#####

 ~X~ 
#NN] 
#####

~~X~|
#NN] 
#####

~~X~|
#NN]|
#####

~~X~|
#NN]~
#####

~~X~|
#NN]~
#####
```

Currently no support for removing blocks, but water seems to be flowing away from the source fairly well. The script gives a demo of a more complex world filling with water.

**Comment by colonelfazackerley** [comment]
Nice. 

How will this look in swing/android? I think that if the water jumps up a whole square it might look odd. how will the graphics be done? particles or states (half full, full, flowing left, ...) with animations like rabbits?

**Comment by tttppp** [comment]
I was thinking of animations to go between states, e.g. always the same 10 frames for river turning into pool. This will probably look a little blocky, but when the water level is static (ie most of the time) it should look ok.

Animating rivers turning into water falls might be tough.

**Comment by colonelfazackerley** [comment]
> Animating rivers turning into water falls might be tough.

That's why I was thinking particles.

**Comment by colonelfazackerley** [comment]
Some thoughts about awkward cases for water.
![awkward_water](https://cloud.githubusercontent.com/assets/9596719/12535432/4566d824-c27b-11e5-998a-065a245cfbd4.png)
If you start some work in  a branch, I would be happy to do pull requests into it.

**Comment by colonelfazackerley** [comment]
A lot of the chars you are using are used already. see

https://github.com/andybalaam/rabbit-escape/blob/master/rabbit-escape-engine/src/rabbitescape/engine/textworld/ChangeRenderer.java

I was going to suggest 'w' for water, but that is taken too. Not sure what to suggest.

**Comment by tttppp** [comment]
I had realised that some of these characters couldn't be the final ones - I'm not even sure they all need distinct characters. I hadn't however realised how tight the remaining character space is getting!  Here's a list of the characters used (according to a script I just wrote):
```
!$%&+,-.:<=>?@DFGHIJKLPRTUWXY[]^_`aefhmsxy|~
```
Also that page didn't seem to contain the symbols for blocks, and bridges - off the top of my head that adds `# \/{}`. Further there's the special character for "see below" `*`!

Maybe all non-deadly water can be `n` and pools can be `N`. If either shares a space with something else then we can use `*` to output the details at the bottom.  (Actually - does the text interface display multiple things in one place with asterisks? I don't remember ever having seen it do that).

**Comment by tttppp** [comment]
I had originally proposed water passing through (or "behind"?) bridges. I don't have a strong opinion on this though. It's another situation to code for, and the animation for partially laid bridges may be more effort than it's worth. On the other hand I can imagine that building aqueducts might be fun!

**Comment by tttppp** [comment]
An example of how I thought water might animate. I've aimed to reuse "animation" where ever possible (e.g. water running down slopes is basically the same as water running off an edge). I'm not sure whether than will work or not. Also I've tried to show what rivers rising within each square might look like.

![water01](https://cloud.githubusercontent.com/assets/7073263/12536800/09adfc9a-c2a8-11e5-9db4-9240c478fbae.png)
![water02](https://cloud.githubusercontent.com/assets/7073263/12536801/0bbb64c8-c2a8-11e5-910f-822edf07053a.png)
![water03](https://cloud.githubusercontent.com/assets/7073263/12536802/0d10cff2-c2a8-11e5-8eef-ff4e9306380c.png)
![water04](https://cloud.githubusercontent.com/assets/7073263/12536803/0e3475aa-c2a8-11e5-8d8d-f9b674f8b2d6.png)
![water05](https://cloud.githubusercontent.com/assets/7073263/12536804/0fc2ca0c-c2a8-11e5-8fff-d48a4d9d8d40.png)
![water06](https://cloud.githubusercontent.com/assets/7073263/12536805/1159e7c4-c2a8-11e5-9e70-7179e6325509.png)
![water07](https://cloud.githubusercontent.com/assets/7073263/12536807/1313b7de-c2a8-11e5-9861-c96c4b755346.png)

**Comment by colonelfazackerley** [comment]
I think the chars in ChangeRenderer are only used for tests. you could do tests another way, but planning early for tests is a good idea.

I dont think 2 levels of fullness (well 3 if you count empty) is enough. Consider a room 20 squares wide next to a tower or well with 1 square internal width. The well will fill fast. There is no way the wide room can fill slowly and smoothly if the state is stored in the way you suggest.

**Comment by tttppp** [comment]
The way I'm suggesting doesn't match physics, but rather is easy (ish) to code and (hopefully) to predict as a player. A well/room fills at exactly the same rate regardless of the width, number of sources feeding it, number of splits in the water since the sources.number

The two levels of filling are because the algorithm (and in theory the player) only needs to worry about "water is passing through this cell" and "rabbits will drown in this cell".

That's my rationale anyway. I can totally understand why we might go for a more complex water model (e.g. A cell can hold at most 10 units of water and sources produce 5 units per time step). But I think it will be complicated, and might not be clear to the player whether their rabbits will drown or not.

**Comment by tttppp** [comment]
Updated python script to show example of a wide room with a well: https://ideone.com/165jny

It looks like:
```
   x      
#        #
#        #
#        #
#### #####
#### #####
#### #####
##########

   *      
#        #
#        #
#        #
#### #####
#### #####
#### #####
##########

   *      
#  |     #
#        #
#        #
#### #####
#### #####
#### #####
##########

   *      
#  |     #
#  |     #
#        #
#### #####
#### #####
#### #####
##########

   *      
#  |     #
#  |     #
#  |     #
#### #####
#### #####
#### #####
##########

   *      
#  |     #
#  |     #
#  ~     #
#### #####
#### #####
#### #####
##########

   *      
#  |     #
#  |     #
# ~~|    #
#### #####
#### #####
#### #####
##########

   *      
#  |     #
#  |     #
#~~~|    #
####|#####
#### #####
#### #####
##########

   *      
#  |     #
#  |     #
#~~~|    #
####|#####
####|#####
#### #####
##########

   *      
#  |     #
#  |     #
#~~~|    #
####|#####
####|#####
####|#####
##########

   *      
#  |     #
#  |     #
#~~~|    #
####|#####
####|#####
####~#####
##########

   *      
#  |     #
#  |     #
#~~~|    #
####|#####
####|#####
####N#####
##########

   *      
#  |     #
#  |     #
#~~~|    #
####|#####
####~#####
####N#####
##########

   *      
#  |     #
#  |     #
#~~~|    #
####|#####
####N#####
####N#####
##########

   *      
#  |     #
#  |     #
#~~~|    #
####~#####
####N#####
####N#####
##########

   *      
#  |     #
#  |     #
#~~~|    #
####N#####
####N#####
####N#####
##########

   *      
#  |     #
#  |     #
#~~~~    #
####N#####
####N#####
####N#####
##########

   *      
#  |     #
#  |     #
#~~~~~   #
####N#####
####N#####
####N#####
##########

   *      
#  |     #
#  |     #
#~~~~~~  #
####N#####
####N#####
####N#####
##########

   *      
#  |     #
#  |     #
#~~~~~~~ #
####N#####
####N#####
####N#####
##########

   *      
#  |     #
#  |     #
#~~~~~~~~#
####N#####
####N#####
####N#####
##########

   *      
#  |     #
#  |     #
#NNNNNNNN#
####N#####
####N#####
####N#####
##########

   *      
#  |     #
#~~~~~~~~#
#NNNNNNNN#
####N#####
####N#####
####N#####
##########

   *      
#  |     #
#NNNNNNNN#
#NNNNNNNN#
####N#####
####N#####
####N#####
##########

   *      
#~~~~~~~~#
#NNNNNNNN#
#NNNNNNNN#
####N#####
####N#####
####N#####
##########

   *      
#NNNNNNNN#
#NNNNNNNN#
#NNNNNNNN#
####N#####
####N#####
####N#####
##########

 ~~X~~~~~ 
#NNNNNNNN#
#NNNNNNNN#
#NNNNNNNN#
####N#####
####N#####
####N#####
##########
```

Sorry for the crazily long comments.

**Comment by andybalaam** [comment]
Loving this thread. Yes,  ChangeDescription characters are a different and less important name space from level files characters. It doesn't matter if we have many to one relationships there e.g. different types of falling all use 'f'.

**Comment by colonelfazackerley** [comment]
The water hits that bulkhead and equal amounts should go each way. Then the huge room and the narrow room  have the height of water increase at the same rate. It's not going to look right.
```
                    *  
                    |  
                       
#                   # #
#                   # #
#                   # #
#                   # #
#                   # #
#######################
```

**Comment by colonelfazackerley** [comment]
I don't mind long posts, BTW.

**Comment by tttppp** [comment]
I don't think it's trivial to model correctly though. Each cell would need a float volume of water (ramps would have half the capacity of empty cells), every split would cause the supplied water in each direction to half. In the real world puncturing a hole in a tube of water causes a different speed of spurt depending on the volume of water above it. Water dropping from a height onto a ramp should splash.

I think we have to limit the modelling of water to something much simpler. Maybe my proposal is too simple though.

**Comment by andybalaam** [comment]
I like your proposal. I think players will accept the world the way it is, and I like the simplicity of the rules that gives nice behaviour.

**Comment by tttppp** [comment]
I just read a couple of pages about water in minecraft. My conclusion is that this isn't an easy problem :-)

I'm not completely sold on my model, for some of the reasons colonelfazackerley has pointed out, but I haven't yet thought of a simple way to improve this. Other things I worry about are water falling speed relative to rabbit falling speed, and relative to river flowing speed. It might not be easy to judge this until we try some java code.

Currently pools fill even if they have no source, so that needs fixing.  This is awkward though, as there are at least two ways a pool can fill. Water falling in from the top, or a source pumping it in from below.

Maybe a different algorithm would solve these problems, eg water produced from the source uses a path finding algorithm and gets added as a block to the current "end" somewhere. This would be more physically accurate (maintaining volume of water) but I'm not sure this would be easy for a player to grasp. It still has the challenge of draining water when blocks are removed.

**Comment by tttppp** [comment]
Another couple of situations (or test cases) I thought of:

* Pools at different heights connected by underground pipes.
* Siphons.

**Comment by colonelfazackerley** [comment]
I had not thought of that. squares that are full will need pressure (depth) values.

**Comment by andybalaam** [comment]
I am inclined to go simpler rather than more complex. I am wondering whether a true cellular automaton would be cool I.e. No interactions between squares over 1 square apart. This would not lead to realistic water flow but why would we want realistic water flow?

**Comment by andybalaam** [comment]
I have been so tired recently I haven't missed any nights' sleep but if/when I do I might try a prototype. On the other hand I might fix some mundane bus :-)

**Comment by GamingInky** [comment]
this looks awesome!

**Comment by tttppp** [comment]
I knocked up a cellular automation version this lunch: https://ideone.com/jedosn

This stores the volume of water in each cell, so water volume is preserved. It has some weird behaviour though - it's not at all easy to guess when different cells will become "pools" (i.e. fatal to rabbits), and it's now possible that not all cells in a row will be at the same depth, so it's usual to get pools next to rivers. Also (and this needs fixing) water doesn't 'level out' around U-bends. I think to fix this we need to store the "pressure" of water in each cell.

Connectivity of cells is really important for this, and since it's the same for rabbits, tokens and water, it will make sense to create a utility that allows access to the connectivity, rather than compute it separately in all places.

I haven't yet perfected the display of my script, so there are temporary floating rivers, but hopefully these examples show roughly how the new algorithm behaves.

[Demo with ramp](http://pastebin.com/mKJYuxWr)

[Room with well](http://pastebin.com/QzWGNEPJ)

[Bulkhead](http://pastebin.com/RGMzprdK)

[Stuck in U-bend](http://pastebin.com/diubL5zm)

**Comment by colonelfazackerley** [comment]
your pastebin'd bits look good. have you thought about how to make it look good in swing/android?

**Comment by tttppp** [comment]
I had a few thoughts, but nothing difinitive.

1. Animate falls, rivers and pools, and variants for river flowing into fall, fall landing on ramp, etc. This requires a lot of animations, but would be in keeping with the current method of animating each cell.

2. Use the volume being constant to create particles as you suggested before. When ware moves between cells then animate the particles moving in that direction. This would be hard to get a clear distinction between rivers, pools and falls (which may cause problems for the player not knowing if water is dangerous or not).

3. Animate falls per cell always looking the same. Animate rivers and falls by showing a blue rectangle at the bottom of the cell. The height of the rectangle could exactly match the water level. This still wouldn't be obvious when the water was at a dangerous level, but it should look fairly smooth.

Even if I get the algorithm working then I'm going to need help with the animation side of things.

**Comment by andybalaam** [comment]
I think we should start with a much simplified version of option 1 looking much like your text. It doesn't have to look real. Then if we design some levels that look really bad we can adjust the animation accordingly.

**Comment by andybalaam** [comment]
I really like the behaviour from your latest pastes. I think u bends can be safely ignored. Unless we think of a level where we need it.

**Comment by andybalaam** [comment]
I'd start without even special animations for slopes.

**Comment by GamingInky** [comment]
I might start doing animations

**Comment by colonelfazackerley** [comment]
I think it might be best to write a programme to generate the animations. A lot will be needed, and water lends itself to coding.

**Comment by andybalaam** [comment]
If there were so many animations that e.g. We doubled the size of the jar I would be concerned.

**Comment by tttppp** [comment]
I spent some time today thinking more about water flow. I think we should aim to get it right first time, because once there are levels with water flowing it will be much harder to change.

I think the U bend problem can be solved by using flood filling to find connected bodies of water, and then giving each node a score based on height it should be possible to try to level the water by moving it along a path to improve the score (reduce the overall gravitational potential energy). Each edge in the network has a maximum capacity for per step, and so iterating will gradually restrict the movement of water that's possible for the time step.

I'm not sure if this makes a lot of sense, but I wanted to make a note in case I forgot.

Ps. I'm excited about the idea of animations. It would be useful to think about what information is needed in order to know how to animate a cell.

**Comment by andybalaam** [comment]
Our automated tests protect us somewhat from changes so long as they are additive rather than change ative.

**Comment by andybalaam** [comment]
But yeah. No hurry, so let's wait until we are happy.

**Comment by tttppp** [comment]
Water is now [flowing round U bends](https://ideone.com/JfPbjL). Unfortunately it's pooling in midair, rather than free-falling.

```
 x #   000000
#  # # 000000
#  # # 000000
#  # # 000000
#    # 000000
###### 000000

 x #   000000
#| # # 050000
#  # # 000000
#  # # 000000
#    # 000000
###### 000000

 x #   000000
#  # # 000000
#N # # 0a0000
#  # # 000000
#    # 000000
###### 000000

 x #   000000
#| # # 050000
#  # # 000000
#N # # 0a0000
#    # 000000
###### 000000

 x #   000000
#  # # 000000
#N # # 0a0000
#  # # 000000
#N   # 0a0000
###### 000000

 x #   000000
#| # # 050000
#  # # 000000
#N # # 0a0000
#~~  # 055000
###### 000000

 x #   000000
#  # # 000000
#N # # 0a0000
#  # # 000000
#~~~ # 066600
###### 000000

 x #   000000
#| # # 050000
#  # # 000000
#N # # 0a0000
#~~~~# 055550
###### 000000

 x #   000000
#  # # 000000
#N # # 0a0000
#  # # 000000
#~~~~# 077770
###### 000000

 x #   000000
#| # # 050000
#  # # 000000
#N # # 0a0000
#~~~~# 077770
###### 000000

 x #   000000
#  # # 000000
#N # # 0a0000
#  # # 000000
#NNNN# 0aaaa0
###### 000000

 x #   000000
#| # # 050000
#  # # 000000
#N # # 0a0000
#NNNN# 0aaaa0
###### 000000

 x #   000000
#  # # 000000
#N # # 0a0000
#~~#~# 033030
#NNNN# 0aaaa0
###### 000000

 x #   000000
#| # # 050000
#  # # 000000
#~~#~# 066060
#NNNN# 0aaaa0
###### 000000

 x #   000000
#  # # 000000
#N # # 0a0000
#~~#~# 066060
#NNNN# 0aaaa0
###### 000000

 x #   000000
#| # # 050000
#  # # 000000
#NN#N# 0aa0a0
#NNNN# 0aaaa0
###### 000000

 x #   000000
#  # # 000000
#N # # 0a0000
#NN#N# 0aa0a0
#NNNN# 0aaaa0
###### 000000

 x #   000000
#| # # 050000
#~~#~# 033030
#NN#N# 0aa0a0
#NNNN# 0aaaa0
###### 000000

 x #   000000
#  # # 000000
#~~#~# 066060
#NN#N# 0aa0a0
#NNNN# 0aaaa0
###### 000000

 x #   000000
#| # # 050000
#~~#~# 066060
#NN#N# 0aa0a0
#NNNN# 0aaaa0
###### 000000
```

**Comment by andybalaam** [comment]
Nearly there

**Comment by GamingInky** [comment]
I was thinking about this and #300 here are to suggestions for interactions between them:

(F is fire)

```
#      #
#    X#
#     |#
#    F#
#####

#      #
#    X#
#     |#
#    ~#
#####,

#       #
#       #
#     F#
#   #N#
######

#       #
#       #
#        #
#    #F#
######

**Comment by colonelfazackerley** [comment]
I don't think I understand why the water source disappears and the extra block appears at the bottom.

**Comment by GamingInky** [comment]
Sorry the copying didn't work I'll fix it...

**Comment by tttppp** [comment]
So I think @gaminginky is suggesting that in some circumstances water is boiled by fire? Presumably this would be when there was a small amount of water in the fire cell?

I'm making progress testing different water algorithms, but I'm not really up to the point of specifying the interactions yet. Maybe this is something we can play with once we have a working water model? (Side note: I read a really interesting article about water modeling in dwarf fortress)

**Comment by GamingInky** [comment]
I think it should be maximum 1 'N'

**Comment by andybalaam** [comment]
I am starting to think water dynamics could be a whole game in itself.  That's not to say we shouldn't do it in Rabbit Escape - I am just wondering whether there will be a spin-off mini game where you have to get the water to the drain or something...

**Comment by tttppp** [comment]
I think I've got something I'm almost happy with.  There's a bug in my code that means occasionally the amount of water in a cell is negative (which you can see by watching the numbers on the right), but other than that, water seems to flow ok.

![basin](https://cloud.githubusercontent.com/assets/7073263/13116958/eabe7642-d595-11e5-84e3-a8fa49461fca.gif)

![bulkhead](https://cloud.githubusercontent.com/assets/7073263/13116961/ed7d05ec-d595-11e5-823d-648a33a42ebb.gif)

![ubend](https://cloud.githubusercontent.com/assets/7073263/13116974/f9692264-d595-11e5-9938-7a27ac473d56.gif)

![pit](https://cloud.githubusercontent.com/assets/7073263/13116973/f9615804-d595-11e5-9329-94c31c0c45cb.gif)

Code here: https://ideone.com/iPJJL1

**Comment by tttppp** [comment]
I forgot to mention that this is cellular automation, whereas the previous version had a routing algorithm in it. Consequently this latest version should be much less memory intensive.

**Comment by colonelfazackerley** [comment]
Very exciting.

The water needs a height difference to spread, but the height difference seems a bit much. Looks a bit like sand piling piling up. 

Might it help if the water had a stepped more often than the rabbits? Perhaps a bad idea: introduce a lot of complexity, like the rabbits falling 2 squares per step.

Need to see it in the game with other things moving to see if it looks OK.

Do you want me to do the animations? I can use similar techniques to the fire.

**Comment by andybalaam** [comment]
Wow! It does look a bit like sand but I think that is likely to provide more puzzle opportunities so looks great to me. Why do the values Flash e.g. Between N and |?

**Comment by andybalaam** [comment]
In the last animation the water in the hole seems to flow back out again sometimes. Is that a bug or a feature of the algorithm?

**Comment by tttppp** [comment]
Oh, and ramps have a maximum capacity of 50, rather than 100.

**Comment by colonelfazackerley** [comment]
For choosing the animation state, I think it will need to know if the cell has water pouring down. Also possibly left or right in the pools.

**Comment by tttppp** [comment]
During each iteration there is a calculation of how much water to move in each of the four directions.  That could be stored against the cell for animation purposes.

In a pool then water will be constantly moving between all adjacent cells. The water level stays roughly the same because it reaches an equilibrium.

**Comment by colonelfazackerley** [comment]
OK, that should work. Let me know when there is an interface to your code that I can work with.

**Comment by andybalaam** [comment]
This is very cool. I would start simple with the animations.

**Comment by andybalaam** [comment]
This is very cool. I would start simple with the animations.

**Comment by tttppp** [comment]
A few more changes, and I'm now happy. Water flows much more consistently and can still flow a reasonably long way around U-bends (although there's probably now a limit).  Unless anyone has any further tests or tweaks then I suggest the next step is Java code!

Well (no more bubbling in the well):
![well](https://cloud.githubusercontent.com/assets/7073263/13187908/2378ee2a-d746-11e5-95b1-41e917ee403d.gif)

U-Bend (water eventually works it's way up the other side):
![ubend](https://cloud.githubusercontent.com/assets/7073263/13187911/28b790bc-d746-11e5-91a2-dea9632ca273.gif)

Bulkhead (this feels very slightly less like sand now):
![bulkhead](https://cloud.githubusercontent.com/assets/7073263/13187919/32008778-d746-11e5-9db1-492b49574d96.gif)

Ramps (I've shown the level and the animated flow separately as it's quite hard to work out what's going on in ASCII otherwise):
![frame_0000](https://cloud.githubusercontent.com/assets/7073263/13187966/656ff972-d746-11e5-8793-c6c5384f0901.png)
![ramps](https://cloud.githubusercontent.com/assets/7073263/13187969/68131ea2-d746-11e5-9794-3127c1cecb42.gif)

Code: https://ideone.com/upD2zy

**Comment by colonelfazackerley** [comment]
Nice.

**Comment by andybalaam** [comment]
That looks ready to me. Fantastic work.

**Comment by tttppp** [comment]
I've made a start on the Java implementation here:

**Comment by colonelfazackerley** [comment]
you used LookupTable2D :)

**Comment by colonelfazackerley** [comment]
A pipe should look like a tap?

**Comment by tttppp** [comment]
I'd thought of it as looking like a circle with water pouring out. Maybe something like this:
http://thumbs.dreamstime.com/z/sewer-pipe-beach-concrete-outlet-nature-32399041.jpg

**Comment by colonelfazackerley** [comment]
Note the convention for test methods. 'generateWaterTable' should be something like 'Generated_water_table_fills_correctly'.

**Comment by colonelfazackerley** [comment]
So we need to take the state of WaterRegion
```
private final Set<Direction> connections;
public int capacity;
public int contents;
private Map<Direction, Integer> flow = new HashMap<>();
```

and note a block in the cell and make an animation for it.

I am thinking it may make sense to not use the rea files, but rather move sprites an amount related to the flow. should end up with less memory used.

**Comment by colonelfazackerley** [comment]
Will pipes be metal? Indestructible by bashers etc?

**Comment by tttppp** [comment]
I was definitely thinking pipes were indestructible. I think they're in the same category as Exits and Entrances.

I'm planning that the water will enter the cell that the pipe is in. This makes connectivity calculations simpler, as pipes don't need to worry about them.

I'm confused as to the difference between calcNewState and step. Which should the water move during? Which should the bunnies drown during?  My guess is that calcNewState should set the flow, and step should move water in line with the flow, but I wanted to check this before I dive in!

**Comment by colonelfazackerley** [comment]
Bunnies should drown in step. You create a new Behaviour called Drowning. It has a behave method (the Rabbit class calls behave in it's step method). In Drowning.behave you call world.changes.killRabbit( rabbit ). See the Falling class.

**Comment by colonelfazackerley** [comment]
> I'm planning that the water will enter the cell that the pipe is in. This makes connectivity calculations simpler, as pipes don't need to worry about them.

OK. I will work with that.

**Comment by colonelfazackerley** [comment]
> I'm confused as to the difference between calcNewState and step. Which should the water move during? Which should the bunnies drown during? My guess is that calcNewState should set the flow, and step should move water in line with the flow, but I wanted to check this before I dive in!

step is called first. This does stuff, so this is when the water or rabbits move. your content values change or rabbit coordinates change. then calcNewState is called to work out what the state will be for the next step. as step is first World.init is called to set things up for the the first step.

**Comment by andybalaam** [comment]
As a side note I'm not happy with the activated step and calcNewState setup but I haven't thought through how to improve it yet.

**Comment by colonelfazackerley** [comment]
@tttppp please could you create a WaterRegionChangeListener in engine like this?
https://github.com/colonelfazackerley/rabbit-escape/commit/1d649bcceb99df466d1217ed644dc4836f18776c

and create a method ```WaterTable.setWaterRegionChangeListener( WaterRegionChangeListener l)```

Then each time a region is added because water spills in, or each time a region is deleted from the table because the water has flowed away, WaterTable uses the methods to update the listener.

**Comment by colonelfazackerley** [comment]
Actually, there is a way that fits in better with the current code. Please can you list the WaterRegions that are new and removed in WorldChanges.

**Comment by colonelfazackerley** [comment]
A suggestion for what to do with git. I pull from andybalaam/master and tttppp/Water regularly. When It's done, I will need to rebase, which is usually a problem. 
when it's done tttppp/Water is merged into andybalaam/master. I pull that and make a fresh branch from it. I then cherry pick my rebased commits from my old Water branch into my new one.

So, no-one else needs to rebase, but I may occassionally ask for some tweaks by @tttppp, to reduce the final merge conflicts.

**Comment by tttppp** [comment]
This sounds fine. I've been deliberately holding off any rebasing or history tweaking so that I don't create extra work for you @colonelfazackerley. If there are larger changes which you want in my branch then please feel free to pull request to it.

I think that at the end we will want to stabilise our work on my branch before we merge anything to andybalaam/master, because otherwise master will be broken ("Invisible water drowns rabbits. Surprise!")

I will have a look at WorldChanges. I haven't addressed bashing/digging/etc yet, but I will need to react to changes in the block table to make updates to the water table.

**Comment by tttppp** [comment]
We could merge tttppp/Water to master earlier if we have some very simple graphics (I'm thinking static blue rectangles), and then we could replace the graphics when we merge cf/Water. Anyway, probably something to think about for the future.

**Comment by tttppp** [comment]
I think changes in the blocks should now result in changes to the water table. @colonelfazackerley  I've updated the WorldChanges class in a way that works, but by listing the coordinates of the changed water regions.

Water currently disappears from these regions, rather than being transferred, but it's a start.

Side note: I did this coding to Peter Rabbit (the CBBC cartoon).

**Comment by colonelfazackerley** [comment]
I was thinking WorldChanges would have

```java
public  final List<>  waterRegionsEmptied = new ArrayList<WaterRegion>();
public  final List<>  waterRegionsStartedFilling = new ArrayList<WaterRegion>();
```

So your list of points have both new and going water regions in? I will need to have a think about wether that works for the WaterRegionRenderer.

**Comment by andybalaam** [comment]
Merging early is good by me. We have lots of tests to check nothing is broken. The later the merge the harder (and therefore the less effective) the review is.

**Comment by tttppp** [comment]
I added some very simple graphics this morning. Falling water looks horrible, but pipes and water are now visible.are To see some water falling then add a P to a space in a level.

Water doesn't yet cause drowning or extinguishing.

**Comment by colonelfazackerley** [comment]
I am now watching CBeebies...

I believe a full cell has 100 water units in. Could this be changed to 1024? For the animations, the cells are a nominal* 32 pixels. 32*32=1024.

This would simplify some thiings for me. Also everything should be a round number (in hex :)

* Can zoom in and out, but in inkscape, and for offsets in rea files it's 32.

**Comment by tttppp** [comment]
We can certainly have 1024 as the standard capacity of a cell. Cells can contain more water than their capacity, although then they are under pressure, and water will leave if possible. Does that cause any problems?

I might experiment with bridges affecting water in the future, which would result in two water regions in the same cell. Maybe water regions should have a shape to make animation easier?

**Comment by colonelfazackerley** [comment]
I have not thought about the pressure issue. How high does the content go?

yes regions probably should have a shape. I am mostly ignoring that for the moment.

**Comment by tttppp** [comment]
With the latest updates, the water model now seems to match the prototype.

@colonelfazackerley There's currently no maximum pressure (although contents is stored in an int, so I guess the answer is currently 32767).  I'd be happy saying `8 x capacity` was the maximum? (Or `16 x capacity` if you want it to be square?)  Fixing a maximum would be a good idea anyway. If a cell contains more than the maximum then is it ok for the excess to just be destroyed?

**Comment by colonelfazackerley** [comment]
The graphics can ignore the excess, I don't mind if you destroy it or not.

I just mean what sort of values are typically reached? Presumably it's deterministic. If you have a water column 5 world tiles deep what's the pressure (excess contents) in the bottom tile?

**Comment by tttppp** [comment]
@colonelfazackerley I did some experiments with a fairly deep U-bend and saw maximum values of about 500 (with the maximum capacity set to 100).  The plan is for a full cell to be lethal to the rabbit - although this is made more complicated by ramps/bridges, as the rabbit's head is in the next cell up.  I've pushed an initial drowning behaviour now.

**Comment by tttppp** [comment]
Ok, done.  I've also added support for extinguishing fire now.  The major thing left is unit tests, as I've been mainly testing by playing so far.  Once I've added some tests then I'll create a pull request.

**Comment by tttppp** [comment]
I've started on tests and found there's a problem exposed by some solutions (eg level 12 easy).

**Comment by colonelfazackerley** [comment]
The water code broke something? Or you happened to notice that 12 easy is broken by something else?

**Comment by tttppp** [comment]
I'm sure it's the water code. It's an out of bounds exception when checking for drowning. I thought maybe that I've got the x and y coordinates mixed up at some point or something.

**Comment by colonelfazackerley** [comment]
This simple level triggers the problem too
```
:solution.1=3
r
 
 
 
 
```

The rabbit is out of bounds, so the world.waterTable.getItemsAt( rabbit.x, yCoordinate ) call is too, I think.

**Comment by andybalaam** [comment]
World record for smallest valid test case?

**Comment by tttppp** [comment]
When I looked at it I thought the Lookup table handled out of bounds for me by allowing a border of one space all the way around. I think the rabbit gets to be two spaces out of bounds, which I didn't think was possible. Anyway, thanks for the test case.

**Comment by tttppp** [comment]
Of course, rabbits fall two spaces per step.

**Comment by tttppp** [comment]
We need to decide whether bridges should block water flow or not. In the original pull request (tttppp/Water) bridges have no impact on water. I've just created a branch (tttppp/BridgesStopFlow) where water is blocked by bridges.  If bridges are made out of "EARTH" then logically I think they should stop water, but other than that I don't have a good reason either way.

Once we pick the interaction we are probably stuck with it, as redirected water breaks some of the solutions to water levels I've created so far.  Some of the levels work either way, so I think we should consider which is best for puzzletunities.

**Comment by colonelfazackerley** [comment]
maybe let water go through bridges. 

anything else will complicate the graphics. would the WaterRegion know if the water in the bridge cell was above or below the bridge?

**Comment by tttppp** [comment]
At the moment a WaterRegion just knows which directions it is (potentially) connected in. It doesnt know anythong about its neighbours, so a space surrounded by square blocks thinks it is connected in all four directions. Due to the blocks in existence at the moment, that information is enough to work out the exact shape of the WaterRegion. A WaterRegion that is only 'connected' upwards looks like a right angled triangle with its hypotenuse at the top of the cell.

We could come up with a better interface of necessary, which would allow us to deal with new blocks.

**Comment by andybalaam** [comment]
Allowing water through bridges sounds simpler and makes sense physically (because there must be a space for the rabbits to get past. I am worried that bridges blocking water is better puzzle wise but am leaning towards simplicity, since this is already more complex than anything else in the game.

**Comment by tttppp** [comment]
Ok, thanks all for your thoughts. It sounds like bridges letting water through is the best option. If we later want some way for rabbits to block water then I suggest we create a new token so rabbits build blocks (or ramps?)

**Comment by colonelfazackerley** [comment]
Building blocks sounds like a good idea.

ramps are necessary, but complicated.

**Comment by andybalaam** [comment]
Further support for water flowing through bridges:
```
( P  )
 (  )
  ()  j
#######
```

If I got the phone ascii right, this rabbit will drown which seems pretty unfair.

**Comment by andybalaam** [comment]
That is supposed to be a jug, which the rabbit can freely walk into, and then down.

**Comment by andybalaam** [comment]
Closing this because it's too massive and takes me minutes to scroll through on my phone.  (And because I merged @tttppp 's pull request).  Feel free to open new issues for further discussions.

Snapshot 0.8.0.7 is here: http://artificialworlds.net/rabbit-escape/snapshots/

**Comment by tttppp** [comment]
This sounds like more proof that bridges aren't actually made of EARTH.

I'm closing this epic thread, and suggest that if there are any remaining issues then they are turned into new tickets. Obviously the water graphics already has a ticket.

**Comment by GamingInky** [comment]
I think bridges aren't very solid I mean there quite thin and slopes are quite thick aren't they
